https://issues.redhat.com/browse/ACM-21429 ClusterInstance field updates (#8260)
open-cluster-management/rhacm-docs: apis/clusterinstance.json.adoc

`SNO`:: {sno}
`HighlyAvailable`:: Multi-node OpenShift|string
|*cpuArchitecture* +

> __optional__|Specify the type of software architecture to use for nodes that do not have an architecture defined. The following software architectures are supported: `multi`, `x86_64`, and `aarch64`. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Reviewer: Can we address the repetition here: `for nodes that do not have an architecture defined` and later `applied to nodes with undefined software architectures`.

> __optional__|Specify the type of software architecture to use for nodes that do not have an architecture defined. The following software architectures are supported: `multi`, `x86_64`, and `aarch64`. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Author: ```suggestion
__optional__|Specify the type of software architecture to use for nodes that do not have an architecture defined. The following software architectures are supported: `multi`, `x86_64`, and `aarch64`.
```

`SNO`:: {sno}
`HighlyAvailable`:: Multi-node OpenShift|string

> |*cpuArchitecture* +

Reviewer 2: cpuArchitecture should be present on both ClusterInstanceSpec (https://github.com/stolostron/siteconfig/blob/release-2.15/api/v1alpha1/clusterinstance_types.go#L420) and NodeSpec (https://github.com/stolostron/siteconfig/blob/release-2.15/api/v1alpha1/clusterinstance_types.go#L192).

If it is not defined at the Node level, then it is inheirited from the ClusterInstance level. If it is not defined at the ClusterInstance level, then it defaults to x86_64

I would also expect the valid values to be present at the ClusterInstance level (https://github.com/stolostron/siteconfig/blob/release-2.15/api/v1alpha1/clusterinstance_types.go#L95), similar to the bootMode doc

Reviewer: @fontivan does `cpuArchitecture` need to be defined at both ClusterInstanceSpec and NodeSpec? If we specify it at the ClusterInstanceSpec and not in the NodeSpec, then the node inherits the ClusterInstanceSpec value right? Setting at the NodeSpec is an override for the specific node. So, it is not required to set both unless there is a mixed architecture deployment consisting of multiple nodes in which case it makes sense to set both. I feel like we are not explaining this clearly here.

Reviewer 2: I still have a few issues with this

- cpuArchitecture is actually optional on both `ClusterInstanceSpec` and `NodeSpec`
- I don't see cpuArchitecture added to the `nodes` section
- I don't see the allowed values added to either clusterinstance (multi, x86, arm) or nodes (x86, arm) section
- I think the verbage around the default value could be better, specifically:
```
The architecture value from `ClusterInstanceSpec` is applied as a default value to nodes with undefined software architectures.
```
and dropping the mention of "level" in favor of:
```
If the `cpuArchitecture` parameter is not defined in `ClusterInstanceSpec`, the default value is `x86_64`
```

This key is added to the host to allow SSH access.|string
|*nodes* +
__required__|Specify the configuration parameters for each node.|<<_rhacm-docs_apis_clusterinstance_jsoncluster_nodes,nodes>> array
|*reference* +
_required_|Specify dependencies or related objects in different namespaces.|array
|*namespace* +

> _required_|Specify the namespace of the reference object.|string

Reviewer: Can you share where you got these fields (viz. `reference` and `namespace`) from?

Author: @sakhoury thanks for your patience! I was on PTO and catching up. Here is the link: https://github.com/stolostron/siteconfig/commit/a804ef8a4b8265f1469861c30a1c57a72489d413@

Reviewer: Thanks @dockerymick! Those are internal source code field names. The user facing fields are described in the CRD. See https://github.com/stolostron/siteconfig/blob/main/config/crd/bases/siteconfig.open-cluster-management.io_clusterinstances.yaml#L321-L337.

Reviewer: So for example,
```yaml
apiVersion: siteconfig.open-cluster-management.io/v1alpha1
kind: ClusterInstance
metadata:
  name: "my-clusterinstance"
  namespace: "my-clusterinstance"
spec:
  ...
  nodes:
    - hostName: "somehost.redhat.com"
      hostRef:
         name: "some-nodename"
         namespace: "some-namespace"
      ...
```

|*templateRefs* +
__required__|Specify the list of the references to cluster-level templates.
A cluster-level template consists of a `ConfigMap` object, in which the keys of the data field represent the kind of the installation manifests.
Cluster-level templates are instantiated once per cluster in the `ClusterInstance` CR.|array
|*hostRef* +

> _optional_|Specify a refence to a `BareMetalHost` resource that is in a different resource.|

Reviewer: I feel like this description is missing key elements from the description in the ClusterInstance CRD:
```
// HostRef is a reference to a BareMetalHost node located in another namespace.
// It is used to link a resource to a specific BareMetalHost instance.
```

> _optional_|Specify a refence to a `BareMetalHost` resource that is in a different resource.|

Author: ```suggestion
_optional_|Specify a reference to a `BareMetalHost` nodes that are in different namespaces. Use `hostRef` to link a resource to a specific `BareMetalHost` instance.|
```

__required__|Specify the list of the references to cluster-level templates.
A cluster-level template consists of a `ConfigMap` object, in which the keys of the data field represent the kind of the installation manifests.
Cluster-level templates are instantiated once per cluster in the `ClusterInstance` CR.|array
|*hostRef* +

> _optional_|Specify a reference to a `BareMetalHost` nodes that are in different namespaces. Use `hostRef` to link a resource to a specific `BareMetalHost` instance.|

Reviewer: ```suggestion
_optional_|Specify a reference to a `BareMetalHost` node that is in a different namespace. Use `hostRef` to link a resource to a specific `BareMetalHost` instance.|
```

|object
|*hostName* +
__required__|Define the host name.|string
|*cpuArchitecture* +

> __optional__|Specify the type of software architecture to use for a specific node that does not have an architecture defined. The `x86_64` and `aarch64` software architectures are supported. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Reviewer: This description here should be tailored for the node and does not need to mention the `ClusterInstanceSpec`.

> __optional__|Specify the type of software architecture to use for a specific node that does not have an architecture defined. The `x86_64` and `aarch64` software architectures are supported. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Author: ```suggestion
__optional__|Specify the type of software architecture to use for a specific node that does not have an architecture defined. The `x86_64` and `aarch64` software architectures are supported.
```

> __optional__|Specify the type of software architecture to use for nodes that do not have an architecture defined. The following software architectures are supported: `multi`, `x86_64`, and `aarch64`. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Reviewer 2: I think this should say 
```Specify the type of software architecture to use for a specific node.```

If I remember right `multi` is only valid on the cluster, not on the node

Author: @fontivan thanks Steven! Updated

> __optional__|Specify the type of software architecture to use for nodes that do not have an architecture defined. The following software architectures are supported: `multi`, `x86_64`, and `aarch64`. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.

Author: ```suggestion
__optional__|Specify the type of software architecture to use for a specific node that does not have an architecture defined. The `x86_64` and `aarch64` software architectures are supported. The architecture from the `ClusterInstanceSpec` is applied to nodes with undefined software architectures.
```

> If the `cpuArchitecture` parameter is not defined in `ClusterInstanceSpec`, the default value is `x86_64`.|string

Reviewer 2: This reads a bit weird

The default value for nodes, if not defined, is the value in clusterinstance.

The default for clusterinstance, if not defined, is x86.

> If the `cpuArchitecture` parameter is not defined in `ClusterInstanceSpec`, the default value is `x86_64`.|string

Author: ```suggestion

If the `cpuArchitecture` parameter is not defined for your nodes, the value in the `ClusterInstance` specification. The default value for the `ClusterInstance` specification is `x86_64`.|string
```

|object
|*hostName* +
__required__|Define the host name.|string
|*cpuArchitecture* +
__optional__|Specify the type of software architecture to use for a specific node that does not have an architecture defined. The `x86_64` and `aarch64` software architectures are supported.

> If the `cpuArchitecture` parameter is not defined for your nodes, the value in the `ClusterInstance` specification. The default value for the `ClusterInstance` specification is `x86_64`.|string

Reviewer 2: The sentence seems cut off here. I would imagine it should read "the value in the `ClusterInstance` specification **is used**"

> If the `cpuArchitecture` parameter is not defined for your nodes, the value in the `ClusterInstance` specification. The default value for the `ClusterInstance` specification is `x86_64`.|string

Author: ```suggestion
If the `cpuArchitecture` parameter is not defined for your nodes, the value in the `ClusterInstance` specification is used. The default value for the `ClusterInstance` specification is `x86_64`.|string
```
